<a href="https://colab.research.google.com/github/lorezolss/RepoDePrueba1/blob/main/Copia_de_ejercicios_supermercado.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ejercicios integrados con dataset de supermercado
Usaremos un solo dataset de ventas para aplicar **lambda**, **map**, **filter**, **reduce** y **comprensiones**.
Cada sección contiene ejercicios de dificultad creciente.


### Dataset inicial

In [ ]:
ventas = [
    {"cliente": "Ana", "producto": "Leche", "categoria": "Lácteos", "precio": 120, "cantidad": 2, "fecha": "2025-01-10"},
    {"cliente": "Luis", "producto": "Pan", "categoria": "Panadería", "precio": 80, "cantidad": 5, "fecha": "2025-01-10"},
    {"cliente": "Sofía", "producto": "Queso", "categoria": "Lácteos", "precio": 300, "cantidad": 1, "fecha": "2025-01-11"},
    {"cliente": "Ana", "producto": "Manzanas", "categoria": "Frutas", "precio": 200, "cantidad": 3, "fecha": "2025-01-12"},
    {"cliente": "Luis", "producto": "Yogur", "categoria": "Lácteos", "precio": 90, "cantidad": 4, "fecha": "2025-01-12"},
    {"cliente": "Marta", "producto": "Carne", "categoria": "Carnes", "precio": 800, "cantidad": 1, "fecha": "2025-01-13"},
    {"cliente": "Sofía", "producto": "Bananas", "categoria": "Frutas", "precio": 150, "cantidad": 6, "fecha": "2025-01-13"},
    {"cliente": "Ana", "producto": "Cereal", "categoria": "Despensa", "precio": 400, "cantidad": 2, "fecha": "2025-01-14"},
]

## Funciones lambda

**Ejercicio:** 1. Crear una lambda que calcule el importe total de una venta (precio * cantidad).

In [ ]:
total = lambda venta: venta["precio"] * venta["cantidad"]
for venta in ventas:
    print(total(venta))
#print(total(ventas[1]))


900
800
800
600
400
360
300
240


**Ejercicio:** 2. Usar una lambda para clasificar los productos como 'barato' si precio < 200 o 'caro' en caso contrario.

In [ ]:
clasificar = lambda venta: "barato" if venta["precio"] < 200 else "caro"
for venta in ventas:
    print(venta["precio"] , clasificar(venta))
#print(clasificar(ventas[2]))

150 barato
800 caro
400 caro
200 caro
80 barato
90 barato
300 caro
120 barato


**Ejercicio:** 3. Ordenar las ventas por fecha usando lambda.

In [ ]:
ventas_x_fecha = lambda venta: venta["fecha"]
ventas.sort(key=ventas_x_fecha)
for venta in ventas:
    print(venta)
#print(ventas)


{'cliente': 'Luis', 'producto': 'Pan', 'categoria': 'Panadería', 'precio': 80, 'cantidad': 5, 'fecha': '2025-01-10'}
{'cliente': 'Ana', 'producto': 'Leche', 'categoria': 'Lácteos', 'precio': 120, 'cantidad': 2, 'fecha': '2025-01-10'}
{'cliente': 'Sofía', 'producto': 'Queso', 'categoria': 'Lácteos', 'precio': 300, 'cantidad': 1, 'fecha': '2025-01-11'}
{'cliente': 'Ana', 'producto': 'Manzanas', 'categoria': 'Frutas', 'precio': 200, 'cantidad': 3, 'fecha': '2025-01-12'}
{'cliente': 'Luis', 'producto': 'Yogur', 'categoria': 'Lácteos', 'precio': 90, 'cantidad': 4, 'fecha': '2025-01-12'}
{'cliente': 'Sofía', 'producto': 'Bananas', 'categoria': 'Frutas', 'precio': 150, 'cantidad': 6, 'fecha': '2025-01-13'}
{'cliente': 'Marta', 'producto': 'Carne', 'categoria': 'Carnes', 'precio': 800, 'cantidad': 1, 'fecha': '2025-01-13'}
{'cliente': 'Ana', 'producto': 'Cereal', 'categoria': 'Despensa', 'precio': 400, 'cantidad': 2, 'fecha': '2025-01-14'}


**Ejercicio:** 4. Ordenar las ventas por importe total de mayor a menor.

In [ ]:
total = lambda venta: venta["precio"] * venta["cantidad"]
ventas.sort(key=total, reverse=True)
#print(ventas)
for venta in ventas:
    print(f'Por : {total(venta)}')
    print(venta)

Por : 900
{'cliente': 'Sofía', 'producto': 'Bananas', 'categoria': 'Frutas', 'precio': 150, 'cantidad': 6, 'fecha': '2025-01-13'}
Por : 800
{'cliente': 'Marta', 'producto': 'Carne', 'categoria': 'Carnes', 'precio': 800, 'cantidad': 1, 'fecha': '2025-01-13'}
Por : 800
{'cliente': 'Ana', 'producto': 'Cereal', 'categoria': 'Despensa', 'precio': 400, 'cantidad': 2, 'fecha': '2025-01-14'}
Por : 600
{'cliente': 'Ana', 'producto': 'Manzanas', 'categoria': 'Frutas', 'precio': 200, 'cantidad': 3, 'fecha': '2025-01-12'}
Por : 400
{'cliente': 'Luis', 'producto': 'Pan', 'categoria': 'Panadería', 'precio': 80, 'cantidad': 5, 'fecha': '2025-01-10'}
Por : 360
{'cliente': 'Luis', 'producto': 'Yogur', 'categoria': 'Lácteos', 'precio': 90, 'cantidad': 4, 'fecha': '2025-01-12'}
Por : 300
{'cliente': 'Sofía', 'producto': 'Queso', 'categoria': 'Lácteos', 'precio': 300, 'cantidad': 1, 'fecha': '2025-01-11'}
Por : 240
{'cliente': 'Ana', 'producto': 'Leche', 'categoria': 'Lácteos', 'precio': 120, 'cantidad': 

**Ejercicio:** 5. Crear una lista de tuplas (cliente, total gastado en esa compra) usando map y lambda.

In [ ]:
cliente_total = list(map(lambda venta: (venta["cliente"], total(venta)), ventas))
print(cliente_total)


[('Sofía', 900), ('Marta', 800), ('Ana', 800), ('Ana', 600), ('Luis', 400), ('Luis', 360), ('Sofía', 300), ('Ana', 240)]


## map()

**Ejercicio:** 1. Generar una lista con los nombres de los clientes en mayúsculas.

In [ ]:
nombres_mayus = list(map(lambda venta: venta["cliente"].upper(), ventas))
print(nombres_mayus)


['SOFÍA', 'MARTA', 'ANA', 'ANA', 'LUIS', 'LUIS', 'SOFÍA', 'ANA']


**Ejercicio:** 2. Calcular con map el importe total de cada venta.

In [ ]:
importe_total = list(map(lambda venta: total(venta), ventas))
print(importe_total)


[900, 800, 800, 600, 400, 360, 300, 240]


**Ejercicio:** 3. Transformar las ventas en strings del tipo: 'Ana compró 2 Leche por $240 el 2025-01-10'.

In [ ]:
desc_venta = list(map(lambda venta: f'{venta["cliente"]} compró {venta["cantidad"]} {venta["producto"]} por ${total(venta)} el {venta["fecha"]}', ventas))
for venta in desc_venta:
    print(venta)
#print(desc_venta)


Sofía compró 6 Bananas por $900 el 2025-01-13
Marta compró 1 Carne por $800 el 2025-01-13
Ana compró 2 Cereal por $800 el 2025-01-14
Ana compró 3 Manzanas por $600 el 2025-01-12
Luis compró 5 Pan por $400 el 2025-01-10
Luis compró 4 Yogur por $360 el 2025-01-12
Sofía compró 1 Queso por $300 el 2025-01-11
Ana compró 2 Leche por $240 el 2025-01-10


**Ejercicio:** 4. Extraer con map solo las categorías de los productos.

In [ ]:
categorias = list(map(lambda venta: venta["categoria"], ventas))
print(categorias)


['Frutas', 'Carnes', 'Despensa', 'Frutas', 'Panadería', 'Lácteos', 'Lácteos', 'Lácteos']


**Ejercicio:** 5. Generar una lista con la longitud de caracteres de cada producto vendido.

In [ ]:
longitud = list(map(lambda venta: len(venta["producto"]), ventas))
print(longitud)


[7, 5, 6, 8, 3, 5, 5, 5]


## filter()

**Ejercicio:** 1. Filtrar todas las ventas de la categoría 'Lácteos'.

In [ ]:
lacteos = list(filter(lambda venta: venta["categoria"] == "Lácteos", ventas))
for venta in lacteos:
    print(venta)
#print(lacteos)



{'cliente': 'Luis', 'producto': 'Yogur', 'categoria': 'Lácteos', 'precio': 90, 'cantidad': 4, 'fecha': '2025-01-12'}
{'cliente': 'Sofía', 'producto': 'Queso', 'categoria': 'Lácteos', 'precio': 300, 'cantidad': 1, 'fecha': '2025-01-11'}
{'cliente': 'Ana', 'producto': 'Leche', 'categoria': 'Lácteos', 'precio': 120, 'cantidad': 2, 'fecha': '2025-01-10'}


**Ejercicio:** 2. Filtrar las ventas realizadas por el cliente 'Ana'.

In [ ]:
ventas_Ana = list(filter(lambda venta: venta["cliente"] == "Ana", ventas))
for venta in ventas_Ana:
    print(venta)
#print(ventas_Ana)


{'cliente': 'Ana', 'producto': 'Cereal', 'categoria': 'Despensa', 'precio': 400, 'cantidad': 2, 'fecha': '2025-01-14'}
{'cliente': 'Ana', 'producto': 'Manzanas', 'categoria': 'Frutas', 'precio': 200, 'cantidad': 3, 'fecha': '2025-01-12'}
{'cliente': 'Ana', 'producto': 'Leche', 'categoria': 'Lácteos', 'precio': 120, 'cantidad': 2, 'fecha': '2025-01-10'}


**Ejercicio:** 3. Filtrar las ventas con importe total > 500.

In [ ]:
vtas_mayores_500 = list(filter(lambda venta: total(venta) > 500, ventas))
for venta in vtas_mayores_500:
    print(venta)
#print(vtas_mayores_500)


{'cliente': 'Sofía', 'producto': 'Bananas', 'categoria': 'Frutas', 'precio': 150, 'cantidad': 6, 'fecha': '2025-01-13'}
{'cliente': 'Marta', 'producto': 'Carne', 'categoria': 'Carnes', 'precio': 800, 'cantidad': 1, 'fecha': '2025-01-13'}
{'cliente': 'Ana', 'producto': 'Cereal', 'categoria': 'Despensa', 'precio': 400, 'cantidad': 2, 'fecha': '2025-01-14'}
{'cliente': 'Ana', 'producto': 'Manzanas', 'categoria': 'Frutas', 'precio': 200, 'cantidad': 3, 'fecha': '2025-01-12'}


**Ejercicio:** 4. Filtrar los productos que se vendieron en cantidad mayor a 3.

In [ ]:
mas_de_3 = list(filter(lambda venta: venta["cantidad"] > 3, ventas))
for venta in mas_de_3:
    print(venta)
#print(mas_de_3)


{'cliente': 'Sofía', 'producto': 'Bananas', 'categoria': 'Frutas', 'precio': 150, 'cantidad': 6, 'fecha': '2025-01-13'}
{'cliente': 'Luis', 'producto': 'Pan', 'categoria': 'Panadería', 'precio': 80, 'cantidad': 5, 'fecha': '2025-01-10'}
{'cliente': 'Luis', 'producto': 'Yogur', 'categoria': 'Lácteos', 'precio': 90, 'cantidad': 4, 'fecha': '2025-01-12'}


**Ejercicio:** 5. Filtrar todas las ventas realizadas en la fecha '2025-01-12'.

In [ ]:
ventas_del_dia = list(filter(lambda venta: venta["fecha"] == "2025-01-12", ventas))
for venta in ventas_del_dia:
    print(venta)
#print(ventas_del_dia)


{'cliente': 'Ana', 'producto': 'Manzanas', 'categoria': 'Frutas', 'precio': 200, 'cantidad': 3, 'fecha': '2025-01-12'}
{'cliente': 'Luis', 'producto': 'Yogur', 'categoria': 'Lácteos', 'precio': 90, 'cantidad': 4, 'fecha': '2025-01-12'}


## reduce()

**Ejercicio:** 1. Calcular la suma total facturada en todas las ventas.

In [ ]:
from functools import reduce
total_facturado = reduce(lambda acc, venta: acc + total(venta), ventas, 0)
print(total_facturado)


4400


**Ejercicio:** 2. Calcular la cantidad total de productos vendidos (sumando todas las cantidades).

In [ ]:
total_prod_vendidos = reduce(lambda acc, venta: acc + venta["cantidad"], ventas, 0)
print(total_prod_vendidos)


24


**Ejercicio:** 3. Encontrar la venta con mayor importe total.

In [ ]:
mayor_venta = reduce(lambda acc, venta: venta if total(venta) > total(acc) else acc, ventas)
print(mayor_venta)


{'cliente': 'Sofía', 'producto': 'Bananas', 'categoria': 'Frutas', 'precio': 150, 'cantidad': 6, 'fecha': '2025-01-13'}


**Ejercicio:** 4. Calcular el gasto total por cliente en un diccionario {cliente: total_gastado}.

In [ ]:
total_x_cliente = reduce(lambda acc, venta: acc | {venta["cliente"]: acc.get(venta["cliente"], 0) + total(venta)}, ventas, {})
print(total_x_cliente)


{'Sofía': 1200, 'Marta': 800, 'Ana': 1640, 'Luis': 760}


**Ejercicio:** 5. Calcular la cantidad de ventas por categoría en un diccionario {'Lácteos': 3, ...}.

In [ ]:
ventas_x_categoria = reduce(lambda acc, venta: acc | {venta["categoria"]: acc.get(venta["categoria"], 0) + 1}, ventas, {})
print(ventas_x_categoria)


{'Frutas': 2, 'Carnes': 1, 'Despensa': 1, 'Panadería': 1, 'Lácteos': 3}


## Comprensiones

**Ejercicio:** 1. Crear una lista con los importes totales de cada venta (precio * cantidad).

In [ ]:
importe_x_venta = [total(venta) for venta in ventas]
print(importe_x_venta)


[900, 800, 800, 600, 400, 360, 300, 240]


**Ejercicio:** 2. Generar una lista con los clientes únicos que compraron algo.

In [ ]:
clientes_unicos = list(set([venta["cliente"] for venta in ventas]))
print(clientes_unicos)


['Ana', 'Marta', 'Sofía', 'Luis']


**Ejercicio:** 3. Crear un diccionario {producto: importe_total} para todos los productos.

In [ ]:
importe_x_producto = {venta["producto"]: total(venta) for venta in ventas}
print(importe_x_producto)


{'Bananas': 900, 'Carne': 800, 'Cereal': 800, 'Manzanas': 600, 'Pan': 400, 'Yogur': 360, 'Queso': 300, 'Leche': 240}


**Ejercicio:** 4. Generar una lista de strings del tipo: 'El cliente Ana compró 3 Manzanas (Frutas)'.

In [ ]:
desc_compra = [f'El cliente {venta["cliente"]} compró {venta["cantidad"]} {venta["producto"]} ({venta["categoria"]})' for venta in ventas]
for venta in desc_compra:
    print(venta)
#print(desc_compra)


El cliente Sofía compró 6 Bananas (Frutas)
El cliente Marta compró 1 Carne (Carnes)
El cliente Ana compró 2 Cereal (Despensa)
El cliente Ana compró 3 Manzanas (Frutas)
El cliente Luis compró 5 Pan (Panadería)
El cliente Luis compró 4 Yogur (Lácteos)
El cliente Sofía compró 1 Queso (Lácteos)
El cliente Ana compró 2 Leche (Lácteos)


**Ejercicio:** 5. Crear un conjunto con todas las categorías únicas vendidas.

In [ ]:
categorias_unicas = set([venta["categoria"] for venta in ventas])
print(categorias_unicas)


{'Lácteos', 'Panadería', 'Despensa', 'Frutas', 'Carnes'}
